### ISSUE TITLE
Display hidden elements in mobile Wikipedia style

### ISSUE THREAD

**Issue by Inbefortus** [kiwix/kiwix-js-windows#198]
Personally, I find reading in mobile style easier than desktop style because the lines are more apart, so you're not distracted by the line underneath. It also does not affect the eyes as much in the long run, but I'm just speaking from personal experience here.

Desktop:

![Screenshot_20210930-184147_Samsung Internet](https://user-images.githubusercontent.com/71934042/135497582-21e99be9-404d-4876-8d69-7674a179e031.jpg)

Mobile:

![Screenshot_20210930-184202_Samsung Internet](https://user-images.githubusercontent.com/71934042/135497545-433f4af3-e8b4-4ec8-81f0-2e6bf1af90af.jpg)

I know I could just use mobile style all the time, but at a great loss: templates.

@Jaifroid What is your take on this?

**Comment by Jaifroid** [comment]
@Inbefortus Two things are possible to solve this: either add an option for wider spacing in Desktop style, as you say, or else have an option to display hidden templates in mobile style.

I kind of like the second option better because it's probably the real reason you use Desktop style, because it is a more complete display. The only issue would be being sure that we can detect hidden content and, more importantly, its class., so that we can override the `display:none` style rule. Ideally we'd want just to be able to select any hidden element, without knowing in advance what we are looking for, and display it. The problem is detecting it based on its calculated style.

The first option may be simpler to achieve technically, but feels like a workaround more than a feature.

**Comment by Inbefortus** [comment]
@Jaifroid Yes, largely solely because of templates. There are several insignificant reasons why I favor desktop style, but honestly, I can say the same about mobile style. Some things are portrayed better on mobile and vice versa, while desktop sometimes displays content that isn't available on mobile and vice versa. 
If both had the same amount of content, it would be a matter of taste how the app portrays content. Since this is not the case, you are kinda forced to switch between desktop and mobile styles to swap certain content for others.

I am also in favor of option 2.

**Comment by Jaifroid** [comment]
I think we would have to iterate over the DOM and then check the computed style of each block-level element:
```
if (document.getComputedStyle(elem).display === 'none') {
    elem.style.display = 'block';
}
```
In a long document, this could be quite slow, but I guess it could be done after the page has rendered initially, or else it could be done only for the currently visible area (like with lazy loaded images).

**Comment by Jaifroid** [comment]
Hey @Inbefortus , could you kindly road test the proposed fix for this issue? To do so, please go to the development PWA at https://kiwix.github.io/kiwix-js-windows/ , browse a bit until you see (in Config) that version 1.7.7-rc1 is ready to load. Then exit the browser window/tab and go back to https://kiwix.github.io/kiwix-js-windows/ (in same browser).

Ensure you have loaded v1.7.7-rc1 correctly -- if it is correctly loaded, you should see a new option "Display hidden block elements" under "Display style" in Configuration. You can also check the version number from Config -> About.

Below shows a page with this option off (left), and with the option turned on (right). Default is off. Note that it only has an effect in mobile Wikipedia style. In Desktop, the hidden elements are always displayed. Elements are displayed with mobile styling (not the desktop styling you may be used to seeing them with).

If you find some elements that are not unhidden with this method, please let me know. It currently examines all `<div>` and `<table>` elements in the document, checks their computed style with `getComputedStyle(element)`, and forces display style `block !important` on the element if its computed style is hidden. If the user turns the option on while viewing a document, then the hidden elements are displayed without a page reload (and for all future articles). If the user turns the option off while viewing a document, then the page is reloaded to return it to its default style. 

![image](https://user-images.githubusercontent.com/4304337/137591233-2e6457ad-06d2-4b1b-84ff-67c68770fc7d.png)

**Comment by Jaifroid** [comment]
I fixed a race condition causing the box to be re-hidden after briefly displaying in some circumstances. This was due to the stylesheet loading fully only after the box had been displayed. The fix is in version 1.7.7-rc2 (ensure you update to this version). It's possible the bug could still occur on old/slow/mobile OS, as there is no way to know when the browser has finished parsing the stylesheets. I have allowed 1 second before forcing display to 'block'. I may need to increase this. (This is the nature of race conditions....)

**Comment by Inbefortus** [comment]
@Jaifroid Great! Very cool that it now shows everything that was hidden like desktop style.


There is only one thing that is a bit annoying (in my opinion). It is the way the templates are displayed. They are partly displayed in an unclear way, from which you don't have a good overview at all, and this makes the page extra long (scrolling wise).


Of course, it would be perfect to display templates exactly like in desktop style, but I think there is not much you can do due to code limits?



Either way, this is definitely a huge improvement, so thank you for your work!

**Comment by Inbefortus** [comment]
@Jaifroid I enabled "Use locally cached display styles" and now everything looks a million times better (thankfully you mentioned this!).


Before:

![Screenshot_20211016-181901_Samsung Internet](https://user-images.githubusercontent.com/71934042/137594873-c43555c7-39dc-4a18-b5ce-c1e5a3584952.jpg)


After:

![Screenshot_20211016-181736_Samsung Internet](https://user-images.githubusercontent.com/71934042/137594885-d6b6023a-d13f-4f80-ad26-067694956877.jpg)


That's exactly how I imagined it.


I still don't mind if you do something that improves clarity, like adding thin lines like you said!

**Comment by Jaifroid** [comment]
Ah yes, I remember when I first enabled caching of stylesheets I was annoyed by some of the differences between Desktop and Mobile style, so I changed some of these things to a more sensible (from my perspective!) default. Like having full-screen-width infoboxes in mobile style, which can be really annoying on a desktop-size display. Hence the locally cached stylesheets are customized for this kind of thing, at least for English-language Wikipedia (annoyingly, each language Wikipedia tends to have its own custom stylesheets...). If you notice things that can be improved, or obvious glitches with these local stylesheets with Wikimedia ZIMs (especially non-English ones), don't hesitate to let me know.
